Userland/Libraries/LibWeb/DOM/DOMTokenList.cpp

```diff
@@ -68,8 +68,8 @@ DOMTokenList::DOMTokenList(Element& associated_element, FlyString associated_att
 {
     m_legacy_platform_object_flags = LegacyPlatformObjectFlags { .supports_indexed_properties = 1 };
 
-    auto value = associated_element.deprecated_get_attribute(m_associated_attribute);
-    associated_attribute_changed(value);
+    if (auto value = associated_element.get_attribute(m_associated_attribute); value.has_value())
```

Comment:
Hm. Is it even correct to check for empty optional here?

The DOM spec says thus:

https://dom.spec.whatwg.org/#ref-for-domtokenlist%E2%91%A0%E2%91%A2

> When a [DOMTokenList](https://dom.spec.whatwg.org/#domtokenlist) object is created, then:
>
> 1. Let element be associated [element](https://dom.spec.whatwg.org/#concept-element).
>
> 2. Let localName be associated attribute’s [local name](https://dom.spec.whatwg.org/#concept-attribute-local-name).
>
> 3. Let value be the result of [getting an attribute value](https://dom.spec.whatwg.org/#concept-element-attributes-get-value) given element and localName.
>
> 4. Run the [attribute change steps](https://dom.spec.whatwg.org/#concept-element-attributes-change-ext) for element, localName, value, value, and null.

And then the result of getting an attribute value is:

https://dom.spec.whatwg.org/#concept-element-attributes-get-value


>To get an attribute value given an [element](https://dom.spec.whatwg.org/#concept-element) element, a string 
 localName, and an optional null or string namespace (default null):
>
> 1. Let attr be the result of [getting an attribute](https://dom.spec.whatwg.org/#concept-element-attributes-get-by-namespace) given namespace, localName, and element.
>
> 2. If attr is null, then return the empty string.
>
> 3. Return attr’s [value](https://dom.spec.whatwg.org/#concept-attribute-value).


I'm pretty sure this means that the signature of `get_attribute` is wrong. It shouldn't return Optional<String>, it should return String, and always return the empty string on "failure"

Reply:
Or is get_attribute correct, and we're just not calling "get attribute value" here?

Reply:
Ah, that should be calling 'get_attribute_value' in that case, not 'get_attribute'. Get attribute value falls back to empty string if missing

Reply:
Probably worth a scrub over the changed spec sections here to make sure there's not more :P. I just saw the first one here and it looked odd to not call the attribute change steps on null attribute

Reply:
Unless I am misunderstanding something, I believe in a lot of cases a `has_attribute` followed by `get_attribute_value` can (and we do) fold them together into `get_attribute` with a `has_value()` check. So in those types of cases "get an attributes value" just ends up being the `.value()` of `get_attribute`. Obviously this isn't a case of that though :P

Reply:
That is, if the spec says "run get_attribute_value" we should do that. if the spec says "if element has an attribute xyz then..." then we should call has_attribute.

Reply:
Yup - we have that here: https://github.com/SerenityOS/serenity/blob/master/Userland/Libraries/LibWeb/DOM/Element.h#L104

It's only called in a few places so far.

(but yeah, for @bplaat, I just chimed in since I knew the answer to the question about `get_attribute_value` after the amount I've been porting this stuff over)

Reply:
> That is, if the spec says "run get_attribute_value" we should do that. if the spec says "if element has an attribute xyz then..." then we should call has_attribute.

Personally  I don't think so, a bit wasteful to do a lookup in the NamedNodeMap twice for an attribute. Just like doing:

```
if (map.find(thing) != thing.end())
   result = *thing.find();
```

Reply:
Okay, I'll change all my changed `get_attribute(...).value_or(String {})` to `get_attribute_value()`. And leave other spec misreads over for a future pr.